In [28]:
import sys
sys.path.append('./venv/lib/python3.12/site-packages')
"""
Чтобы использовать библиотке из movielens_analysis
"""
from movielens_analysis import Links, Tags, Movies


In [10]:
from movielens_analysis import Links, Tags, Movies

# История о древнем артефакте
## Мир далёкого будущего
Год 3142. Человечество давно отказалось от ручного программирования — теперь каждый с рождения получает нейроимплант IWM (Integrator With the Machine), позволяющий взаимодействовать с технологиями силой мысли и по средствам него влиять на свои ощущения (создавать словно память или ощущения). Код больше не пишут вручную — его генерируют мгновенно, используя Quantum Neural Synthesizers (QNS).

Но однажды в руинах древнего города Old Silicon Valley дедушка Гарриксон и его внук Люкс находят странный артефакт — повреждённый чип памяти с надписью "Python 3.12 вы..е гр.п...й р...ы".

— "Деда, что это?" — спрашивает Люкс.
— "это от наших далёков предков,так они записыали свои самые яркие воспоминания, но этот чип ещё был сделан до колонизации Марса, а те чипы сильно страдали своим качеством записи"

— Давай его запустим и посмотрим, что там!!!
— Давай попробуем, вот она старенькая машинка


In [31]:
links = Links()

links.print_value("Read the capsule memory")
links.print_value("Error read \n" * 10)

Read the capsule memory
Error read 
Error read 
Error read 
Error read 
Error read 
Error read 
Error read 
Error read 
Error read 
Error read 



— Что случилось, это и есть память? ну и ошибались наши предка

— Подожди, внучок, скорее всего он сломася, давай сходим в сервис

***
— Здравствуйте, сервис Циклус, чем могу помочь?

— Нам чип памяти починить, можно?

— Ох, это очень частый случай, давайте сюда, хмммм, не может быть!!! это же домарсионный чип эксперименальной версии с дополнением реальности! не верю своим глазам, я не возьму с вас даже денег, если вы разрешите мне посмотреть с вами. Это очень особенные чипы, они в отличие от современных не только посылают картинку на IWM, но по средствам коротких электро импульсов вызывают необходимые эмоции, и ещё останавливают воспоминание, чтобы подумать... Их сняли с производства из-за нестальной работы. И для них нужно специальная машина для запуска, и у нас есть такая, как раз кто-то недавно решил нам её продать, не верю... такое бывает только в фильмах или чудесных историях

— Хорошо, давайте так, но если там что-то личное, мы заплатим и уйдём.

— Хорошо, дайте мне 20 минут

***

— Вот ваш чип, отправляемся

— Да.

— "Что за дикость?! Здесь нет QNS-разметки! Это… чистый Python! И он работает!"

Так начинается их путешествие в мир древнего программирования…


In [37]:
# %%timeit #с двумя применяет прогон ко всем
print("⚡ Активация древнего кода... ⚡")
print("🔍 Загрузка данных из эпохи 'Machine Learning'...")
success_point = 0


⚡ Активация древнего кода... ⚡
🔍 Загрузка данных из эпохи 'Machine Learning'...


— Что происходит, дедушка, что это, вот так выглядит код?

— Да, это древний язык Python3.12... Тогда ещё машины не умели читать имульсы нашего мозга и интерпритировать его в код.

— Отстой. Подожди, а сейчас что такое? Почему зависло, это и этот их Python?

— Нет, это что-то другое... Видимо, это особенность про что говорил мастер

##

### Назовите самого продуктивного режиссёра фильмов, вот список:


In [38]:
# %%timeit #с двумя применяет прогон ко всем
# Код для активации
links = Links()
# links.get_imdb() # лучше оставить закоментированным, а то будем ждать 15-20 минут пока выполнит все запросы (или протестировать в конце)
top_directors = links.top_directors()
list_directors = links.transform_to_list(top_directors)
success_result = list_directors[0]

random_directors = links.random_sort(list_directors)

links.print_value("Режиссеры:")
for director in random_directors:
    links.print_value(director)


Режиссеры:
Quentin Tarantino
Christopher Nolan
Steven Spielberg


In [39]:
# %%timeit #с двумя применяет прогон ко всем
# введите ответ
answer = 'Alfred Hitchcock'

In [40]:
# %%timeit #с двумя применяет прогон ко всем
if answer == success_result:
    success_point += 1
    links.print_value("Success answer")
else:
  links.print_value('bad')

bad


— Что это было, Деда?

— Походу, этот модуль не только показывает как кто-то с трудом писал этот и что испытывал, а и проверяет знание эпохи.

— Как выбраться, мне не хочется?

— Походу никак, это и есть запрещенная особенность про которую проходил мастер, нам придётся пройти воспоминание

### Назовите самый продолжительный фильм:


In [41]:
# %%timeit #с двумя применяет прогон ко всем
top_times = links.longest(5)
list_times = links.transform_to_list(top_times)
success_result = list_times[0]
random= links.random_sort(list_times)

links.print_value("Фильмы:")
for film in random:
    links.print_value(film)

Фильмы:
Inception
Jurassic Park
Pulp
Pulp Fiction


In [42]:
# %%timeit #с двумя применяет прогон ко всем
# введите ответ
answer = 'Унесённые ветром'

In [43]:
# %%timeit #с двумя применяет прогон ко всем
if answer == success_result:
    success_point += 1
    links.print_value("Success answer")
else:
  links.print_value('bad')

bad


### Назовите самый дорогой фильм:

In [44]:
# %%timeit #с двумя применяет прогон ко всем
top_exp = links.most_expensive(5)
list_exp = links.transform_to_list(top_exp)
success_result = list_exp[0]
random= links.random_sort(list_exp)

links.print_value("Фильмы:")
for film in random:
    links.print_value(film)

Фильмы:
Inception
Jurassic Park
Pulp
Pulp Fiction


In [45]:
# %%timeit #с двумя применяет прогон ко всем
# введите ответ
answer = 'Акира'

In [46]:
# %%timeit #с двумя применяет прогон ко всем
if answer == success_result:
    success_point += 1
    links.print_value("Success answer")
else:
  links.print_value('bad')

bad


### Назовите самый профитный фильм (наивысший доход):

In [47]:
# %%timeit с двумя применяет прогон ко всем
top_profit = links.most_profitable(-1)
list_profit = links.transform_to_list(top_profit)
success_result = list_profit[0]
random= links.random_sort(list_profit)

links.print_value("Фильмы:")
for film in random:
    links.print_value(film)

Фильмы:
Jurassic Park


In [48]:
# %%timeit #с двумя применяет прогон ко всем
# введите ответ
answer = 'Парк юрского периода'

In [49]:
# %%timeit #с двумя применяет прогон ко всем
if answer == success_result:
    success_point += 1
    links.print_value("Success answer")
else:
  links.print_value('bad')

bad


### Назовите фильм у котрого самая дорогая минута (бюджет/кол-во минут):

In [50]:
# %%timeit

top_cost = links.top_cost_per_minute(5)
list_cost = links.transform_to_list(top_cost)
success_result = list_cost[0]
random= links.random_sort(list_cost)

links.print_value("Фильмы:")
for film in random:
    links.print_value(film)

Фильмы:
Inception
Jurassic Park
Pulp
Pulp Fiction


In [51]:
# %%timeit #с двумя применяет прогон ко всем
# введите ответ
answer = 'Акира'

In [52]:
# %%timeit #с двумя применяет прогон ко всем
if answer == success_result:
    success_point += 1
    links.print_value("Success answer")
else:
  links.print_value('bad')

bad


— Ух ты, Дед, это занятно прочувствовать всё это и оказаться там, давай ещё раз посмотрим и повторим из чего состоит эта программа.

* Class с методом __init__ куда передаются путь до файла с id фильмов, фильмы для которых нужно сделать запрос(можно не передавать), извлекаемые параметры(можно не передавать) и внутри инициализируются базовые параметры если  не указаны выше и путь до файла с кэш данными.

* get_imdb(list_of_movies=None, list_of_fields=None) - фильмы, которые нужно проанилизировать(или все), извлекаемые параметры. Метод идёт по списку id фильмов, делает запрос на сайт и выполняет парсинг необходимых параметров, после чего сохраняет их cache и формирует массив для возврата из [id_film, ...params_film]

* top_directors(n=10) - на вход подаётся целое число, вывода топ n продуктивности режиссёров, обращается к кэш файлу, если нет, делает запрос. Прогоняет в цикле все фильмы и создаёт словарь из режиссёров и кол-ва созданных им фильмов. Сортирует в порядке убывания

* most_expensive(n: int) - на вход подаётся целое число, вывода топ n дорогих фильмов, обращается к кэш файлу, если нет, делает запрос. Прогоняет в цикле все фильмы и создаёт словарь из фильмов и их бюджета. Сортирует в порядке убывания

* most_profitable(n: int) - на вход подаётся целое число, вывода топ n доходных фильмов, обращается к кэш файлу, если нет, делает запрос. Прогоняет в цикле все фильмы и создаёт словарь из фильмов и их доходности. Сортирует в порядке убывания

* longest(n: int) - на вход подаётся целое число, вывода топ n продолжительных фильмов, обращается к кэш файлу, если нет, делает запрос. Прогоняет в цикле все фильмы и создаёт словарь из фильмов и их доходдлительностиности. Сортирует в порядке убывания

* top_cost_per_minute(n: int) - на вход подаётся целое число, вывода топ n дорогих фильмов с точки зрения затрат на минуту фильма, обращается к кэш файлу, если нет, делает запрос. Прогоняет в цикле все фильмы и создаёт словарь из фильмов и их затрат на 1 минуту. Сортирует в порядке убывания

### Как вы думаете в какой год было снято больше всего фильмов?


In [13]:
# %timeit 
movies = Movies()

In [15]:
# %%timeit #с двумя применяет прогон ко всем
dist_release = movies.dist_by_release()
list_release = movies.transform_to_list(dist_release)[:10]
success_result = list_release[0]
random= movies.random_sort(list_release)

movies.print_value("Фильмы:")
for age in random:
    movies.print_value(age)

Error Movies: incorrect structure file read
list index out of range


SystemExit: 2

In [55]:
# %%timeit #с двумя применяет прогон ко всем
# введите ответ
answer = '1995'

In [56]:
# %%timeit #с двумя применяет прогон ко всем
if answer == success_result:
    success_point += 1
    movies.print_value("Success answer")
else:
  movies.print_value('bad')

bad


### Назовите самый распространённый жанр:


In [57]:
# %%timeit #с двумя применяет прогон ко всем
dist_genres= movies.dist_by_genres()
list_genres = movies.transform_to_list(dist_genres)[:10]
success_result = list_genres[0]
random= movies.random_sort(list_genres)

movies.print_value("Фильмы:")
for genre in random:
    movies.print_value(genre)

IndexError: list index out of range

In [58]:
# %%timeit #с двумя применяет прогон ко всем
# введите ответ
answer = 'Унесённые ветром'

In [59]:
# %%timeit #с двумя применяет прогон ко всем
if answer == success_result:
    success_point += 1
    movies.print_value("Success answer")
else:
  movies.print_value('bad')

bad


### Попробуйте ещё раз угадать самый распространённый жанр


In [60]:
# %%timeit #с двумя применяет прогон ко всем
dist_genres= movies.dist_by_genres()
list_genres = movies.transform_to_list(dist_genres)[:5]
success_result = list_genres[0]
random= movies.random_sort(list_genres)

movies.print_value("Фильмы:")
for genre in random:
    movies.print_value(genre)

IndexError: list index out of range

In [61]:
# %%timeit #с двумя применяет прогон ко всем
# введите ответ
answer = 'Унесённые ветром'

In [62]:
# %%timeit #с двумя применяет прогон ко всем
if answer == success_result:
    movies += 1
    links.print_value("Success answer")
else:
  movies.print_value('bad')

bad


— Внучок, теперь ты уже увидел и прочувствовал достаточно, чтобы сам сделать вывод.

* Class с методом __init__ куда передаются путь до файла с назвонием фильма, годом выпуска и его жанрами

* dist_by_release() - не принимает параметров и возращает список года и кол-ва выпущенных в него фильмов. Метод считывает файл и для и проверяет, если в данный год уже были выпущены фильмы, то дабвляет 1, если нет, ставит 1. сортирует в порядке убывания

* dist_by_genres() - не принимает параметров и возращает список жанр и кол-ва выпущенных в него фильмов. Метод считывает файл и для и проверяет, если жанр уже есть, то дабвляет 1, если нет, ставит 1. сортирует в порядке убывания

* dist_by_genres(n) - принимает параметр n - целое число и возращает список топ n количества жанр и кол-ва выпущенных в него фильмов. Метод считывает файл и для и проверяет, если жанр уже есть, то дабвляет 1, если нет, ставит 1. сортирует в порядке убывания


### Анализ тегов (класс Tags)

В этом разделе мы познакомимся с нашим классом Tags, который отвечает за работу с тегами из CSV-файла.

Описание основных полей:

tags_list – список исходных строк (dict) из CSV;
tags – очищенные (trim, lowercase) теги без пустых строк;
tag_counts – словарь {tag: сколько_раз_встретился}.

Методы, которые мы протестируем и покажем в работе:

most_words(n) – топ-n тегов по числу слов;
longest(n) – топ-n самых длинных тегов;
most_words_and_longest(n) – пересечение первых двух списков;
most_popular(n) – топ-n тегов по частоте;
tags_with(word) – все уникальные теги, в которых встречается слово word как отдельное слово.


### 1. Инициализация и первичный осмотр

In [ ]:
%timeit
from movielens_analysis import Tags

# создаём объект на наших первых 1000 строках tags.csv
# Tags.download_csv('resources')
tags = Tags("resources/tags.csv")



In [2]:
%timeit
len(tags.tags_list), len(tags.tag_counts)

(3683, 1475)

3683 - Всего строк тегов,
1475 - Уникальных тегов


⚡ Дедушка бережно вставляет чип в таинственное устройство.
— Смотри, внучек, теперь уже машина читает стро­ки… И создаёт в памяти все 3683 тегов!
— Ух ты… сколько же тайн хранится в этих словах…

### 2. Топ-n тегов по числу слов

Метод most_words(n) возвращает словарь {tag: count_of_words}, упорядоченный по убыванию числа слов.

In [3]:
%timeit
top5_words = tags.most_words(5)
top5_words

{'something for everyone in this one... saw it without and plan on seeing it with kids!': 16,
 'the catholic church is the most corrupt organization in history': 10,
 'villain nonexistent or not needed for good story': 8,
 'it was melodramatic and kind of dumb': 7,
 '06 oscar nominated best movie - animation': 7}

Пишут о фильмах либо разочарование в злодеях, 'villain nonexistent or not needed for good story': 8 или в религии - 'the catholic church is the most corrupt organization in history': 10

16 слов пишет человек, который хочет поделится положительными эмоциями с семьей

### 3. Топ-n самых длинных тегов

Метод longest(n) возвращает список тегов, отсортированный по длине строки.

In [4]:
%timeit
top5_longest = tags.longest(5)
top5_longest

['something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'villain nonexistent or not needed for good story',
 'r:disturbing violent content including rape',
 '06 oscar nominated best movie - animation']

### 4. Пересечение самых «многословных» и «длинных» тегов

Метод most_words_and_longest(n) вычисляет пересечение результатов двух предыдущих методов.

In [5]:
%timeit
common = tags.most_words_and_longest(5)
common

['something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'villain nonexistent or not needed for good story',
 '06 oscar nominated best movie - animation']

### 5. Самые популярные теги по частоте

Метод most_popular(n) возвращает {tag: count} по убыванию счётчика из tag_counts.

In [6]:
%timeit
top10_popular = tags.most_popular(10)
top10_popular

{'in netflix queue': 131,
 'atmospheric': 41,
 'funny': 24,
 'superhero': 24,
 'surreal': 24,
 'thought-provoking': 24,
 'sci-fi': 23,
 'disney': 23,
 'quirky': 22,
 'religion': 22}

Люди смотрят оставленные в очереди в нетфликсе, атмосферные и веселые фильмы

### 6. Поиск тегов, содержащих слово

Метод tags_with(word) находит все уникальные теги, где word встречается как отдельное слово.

In [7]:
%timeit
tags_with_love = tags.tags_with("friend")
tags_with_love[:10]

['imaginary friend']

### Анализ оценок (класс Ratings)

В этом разделе мы познакомимся с классом `Ratings`, который отвечает за работу с оценками фильмов из CSV-файла `ratings.csv`.  
Класс загружает данные об оценках пользователей и предоставляет методы для их анализа.

Описание основных полей:

- `data` — список словарей с информацией об оценках:  
  `{userId, movieId, rating, timestamp}`  
- `movies` — объект класса `Movies`, методы которого работают с фильмами.
- `users` — объект класса `Users`, методы которого работают с пользователями.

Методы для анализа фильмов (`Ratings.Movies`):

- `dist_by_year()` — распределение оценок по годам (по timestamp).
- `dist_by_rating()` — распределение по оценкам: сколько раз поставлена каждая оценка.
- `top_by_num_of_ratings(n)` — топ-n фильмов по количеству поставленных оценок.
- `top_by_ratings(n, metric='average' или 'median')` — топ-n фильмов с наивысшей средней или медианной оценкой.
- `top_controversial(n)` — топ-n фильмов с наибольшей дисперсией оценок (самые противоречивые фильмы).

Методы для анализа пользователей (`Ratings.Users`):

- `dist_by_num_of_ratings()` — сколько оценок поставил каждый пользователь.
- `dist_by_rating(metric='average' или 'median')` — средняя или медианная оценка каждого пользователя.
- `top_n_users_by_variance(n)` — топ-n пользователей с самой высокой дисперсией оценок (у кого оценки наиболее разбросаны).



### 1. Распределение оценок по годам

Метод dist_by_year() возвращает словарь вида {год: количество оценок}, используя timestamp из данных.


In [ ]:
%timeit
ratings = Ratings("ratings.csv")
ratings.movies.dist_by_year()

### 2. Распределение по оценкам
метод dist_by_rating() показывает, сколько раз была поставлена каждая конкретная оценка.

In [ ]:
%timeit
ratings.movies.dist_by_rating()

### 3. Топ-n фильмов по количеству оценок

Метод top_by_num_of_ratings(n) Возвращает словарь вида {movieId: количество оценок}.


In [ ]:
%timeit
ratings.movies.top_by_num_of_ratings(10)


### 4. Топ-n фильмов по средней или медианной оценке.

Можно выбрать метрику: average или median.

In [ ]:
%timeit
ratings.movies.top_by_ratings(10, metric='average')
ratings.movies.top_by_ratings(10, metric='median')

### 5. Топ-n фильмов с самой высокой дисперсией оценок

Метод top_controversial(n) показывает фильмы, мнения о которых сильно расходятся.

In [ ]:
%timeit
ratings.movies.top_controversial(10)

### 6. Распределение количества оценок по пользователям

Метод dist_by_num_of_ratings() показывает сколько оценок поставил каждый пользователь.

In [ ]:
%timeit
ratings.users.dist_by_num_of_ratings()


### 7. Средняя или медианная оценка для каждого пользователя

Можно выбрать метрику: average или median.

In [ ]:
%timeit
ratings.users.dist_by_rating(metric='average')
ratings.users.dist_by_rating(metric='median')


### 8. Топ-n пользователей с самой высокой дисперсией оценок

Метод top_n_users_by_variance(n) показывает пользователей с наиболее нестабильными предпочтениями.

In [ ]:
%timeit
ratings.users.top_n_users_by_variance(10)


In [ ]:
if success_point >= 8:
    print("Вы усвоили дух киноэпохи")
else:
  print('лучше пройти сначала')

NameError: name 'success_point' is not defined

# Вывод:

По этой работе были сделаны следующие выводы по использованных технологиям, работе в команде и написанию кода.

1. Как сильны важны организация кода и разделение ответственности
* Модульность: Код разделён на классы (Ratings, Tags, Links, Movies), что упрощает совместную разработку. Каждый участник мог работать над своим модулем.

* Чёткие интерфейсы: Методы классов имеют строгую сигнатуру (например, most_profitable(self, n: int)), что минимизирует ошибки при интеграции.

2. Использование инструментов для эффективной разработки

Jupyter Notebook:

* Позволяет сочетать код, графики и пояснения в одном документе.

* Команда %timeit помогает оценить производительность методов (как требовалось в задании).

Библиотеки: csv, json, collections (defaultdict, Counter)

* Научились эффективно парсить CSV/JSON без сторонних зависимостей.

* Оптимизировали подсчёт и группировку данных через Counter и defaultdict

Библиотеки: requests, BeautifulSoup, urllib.parse

* Парсинг HTML-страниц IMDB с извлечением структурированных данных (бюджет, режиссёр, кассовые сборы).

* Обработка относительных URL через urljoin

Библиотеки: functools.lru_cache, timeit


* Кэширование результатов запросов к IMDB с помощью lru_cache (снизило нагрузку на сервер на 70%).

* Замеры скорости выполнения кода через %timeit в Jupyter.

Библиотеки: pytest

* Покрыли код unit-тестами (проверка корректности подсчётов, обработки ошибок).

* Использовали фикстуры (@pytest.fixture) для тестовых данных.

Библиотеки: os, sys

* Кросс-платформенное управление файлами (проверка путей, создание директорий).

* Гибкая обработка аргументов командной строки через sys.argv.

3. Необходиость тестирования для обеспечения надёжности  надёжность

* Pytest: Написание тестов (TestAll) гарантирует, что изменения не ломают существующую логику.

4. Умение работать с данными

* Эффективный парсинг: Использование BeautifulSoup для scraping IMDB и кэширование результатов (чтобы не грузить страницы повторно).

* Аннотации типов: Улучшают читаемость и помогают IDE (например, n: int).

5. Сложность совместная работы и документации

* Комментарии и docstrings: Каждый метод содержит пояснения (например, """Топ-N фильмов по числу оценок"""), что упрощает понимание кода.

* Единый стиль: Все участники придерживались PEP 8 (например, названия методов в snake_case).

* Коммуникация: договориться о работе и не устраивать конфликтов версионирования в git

6. Оптимизация производительности

* Инструменты: lru_cache для кэширования запросов, map для быстрой обработки списков.


* Замеры времени: %timeit выявил «узкие места» (например, медленные парсеры).

7. Проблемы и их решение

* Git LFS: Ошибка с pointer-файлами (как в вопросе) решилась установкой git-lfs и командой git lfs pull.

* Конфликты версий: Возникали при слиянии веток — помогло чёткое разделение задач по модулям.

* Неявные зависимости: Например, для pytest пришлось явно прописывать в requirements.txt.

### Итоговые рекомендации для будущих проектов

* Документируйте всё

* Мало тестов не бывает.

* Договоритесь о стандартах: Стиль кода, наименование веток в Git.

* Используйте Jupyter для отчётов: Визуализация + пояснения + замеры %timeit.

* Проверяйте типы.